### PR TITLE
Allow the boundary conditions of scalar fields to be coupled

### PR DESCRIPTION
**Pull request by cianwilson**
This PR allows the boundary values or radial gradients of scalar fields (including the thermal variable and active scalar fields) to be linearly coupled together.

The new functionality is documented in the "Under Development" section of the User Guide.

The most pressing point for review is that this changes the number of boundary condition types from two (Dirichlet and Neumann) to four (Dirichlet, Neumann, coupled Dirichlet and coupled Neumann).  Combined with the fact that, in an effort to not add too many new options, `T_top`, `T_bottom`, `dTdr_top`, `dTdr_bottom`, `chi_a_top`, `chi_a_bottom`, `dchidr_a_top`, `dchidr_a_bottom`, and their `_file` equivalents have been overloaded, this means that the previous defaults (e.g. `fix_Tvar_top=.true.` with `T_top=0`) have had to be taken away.  The default behavior now checks for multiply set boundary conditions and throws an error, or, if no boundary conditions are set, initializes the options to the old defaults while issuing a warning.

Otherwise I don't think this intersects with existing behavior.

A test for constant bcs has been added in `tests/coupled_bcs`.  Future work should add one for spatially varying conditions.

### CODE REVIEW

File: src/Physics/Sphere_Linear_Terms.F90 [hunk 1]
```diff
@@ -573,6 +621,32 @@ Subroutine Set_Boundary_Conditions(mode_ind)
...
+                  end do
+                  samp = -chi_T_coeff_top(i)
+                  Call Load_BC(lp,r,chiaeq(i),tvar,samp,0)
+                  samp = -chi_dTdr_coeff_top(i)
+                  Call Load_BC(lp,r,chiaeq(i),tvar,samp,1)
+              Endif
+              If (couple_dchidr_a_top(i)) Then
+                  do j = 1, n_active_scalars
+                      samp = -dchidr_chi_coeff_top(i,j)
+                      if (i==j) samp = 1.0
```
Comment (cianwilson) [inline src/Physics/Sphere_Linear_Terms.F90:640]: Move this line to the dchidr

File: src/Physics/Sphere_Linear_Terms.F90 [hunk 2]
```diff
@@ -581,8 +655,35 @@ Subroutine Set_Boundary_Conditions(mode_ind)
...
+                  end do
+                  samp = -chi_T_coeff_bottom(i)
+                  Call Load_BC(lp,r,chiaeq(i),tvar,samp,0)
+                  samp = -chi_dTdr_coeff_bottom(i)
+                  Call Load_BC(lp,r,chiaeq(i),tvar,samp,1)
+              Endif
+              If (couple_dchidr_a_bottom(i)) Then
+                  do j = 1, n_active_scalars
+                      samp = -dchidr_chi_coeff_bottom(i,j)
+                      if (i==j) samp = 1.0
```
Comment (cianwilson) [inline src/Physics/Sphere_Linear_Terms.F90:674]: Needs to be moved too

File: src/Physics/BoundaryConditions.F90 [hunk 3]
```diff
@@ -63,11 +73,41 @@ Module BoundaryConditions
...
+    Real*8  :: T_dchidr_coeff_bottom(1:n_scalar_max)  = 0.0d0
+    Real*8  :: T_dTdr_coeff_bottom                    = 0.0d0
+    Real*8  :: dTdr_chi_coeff_top(1:n_scalar_max)     = 0.0d0
+    Real*8  :: dTdr_dchidr_coeff_top(1:n_scalar_max)  = 0.0d0
+    Real*8  :: dTdr_T_coeff_top                       = 0.0d0
+    Real*8  :: dTdr_chi_coeff_bottom(1:n_scalar_max)     = 0.0d0
+    Real*8  :: dTdr_dchidr_coeff_bottom(1:n_scalar_max)  = 0.0d0
+    Real*8  :: dTdr_T_coeff_bottom                       = 0.0d0
+
     Real*8  :: chi_a_Bottom(1:n_scalar_max)     = 1.0d0
```
Comment (cianwilson) [inline src/Physics/BoundaryConditions.F90:89]: missed this

**Comment by cianwilson** [comment]
Need to set this up to restore the old defaults if no other bc is set

**Comment by cianwilson** [comment]
Just rebased this branch on top of main

**Review by feathern — APPROVED** [review]
Awesome. The docs look good, and all checks passed.   Approving and merging now.
